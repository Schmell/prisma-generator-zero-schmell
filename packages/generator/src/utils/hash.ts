import type { DMMF } from "@prisma/generator-helper";
import { createHash } from "crypto";

export function generateSchemaHash(models: DMMF.Model[], enums: DMMF.DatamodelEnum[]): string {
  const hash = createHash("sha256");

  // Only hash the structural elements that affect the schema
  const schemaStructure = {
    // Sort models by name for consistency
    models: [...models]
      .sort((a, b) => a.name.localeCompare(b.name))
      .map((model) => ({
        name: model.name,
        dbName: model.dbName,
        // Sort fields by name for consistency
        fields: [...model.fields]
          .sort((a, b) => a.name.localeCompare(b.name))
          .map((f) => ({
            name: f.name,
            type: f.type,
            isRequired: f.isRequired,
            isList: f.isList,
            relationName: f.relationName,
            relationFromFields: f.relationFromFields ? [...f.relationFromFields].sort() : [],
            relationToFields: f.relationToFields ? [...f.relationToFields].sort() : [],
            default: f.default,
            unique: f.isUnique,
          })),
        primaryKey: model.primaryKey,
        uniqueFields: model.uniqueFields ? [...model.uniqueFields].sort() : [],
        uniqueIndexes: model.uniqueIndexes,
      })),
    // Sort enums by name for consistency
    enums: [...enums]
      .sort((a, b) => a.name.localeCompare(b.name))
      .map((enumType) => ({
        name: enumType.name,
        values: [...enumType.values]
          .sort((a, b) => a.name.localeCompare(b.name))
          .map((v) => ({
            name: v.name,
            dbName: v.dbName,
          })),
      })),
  };

  hash.update(JSON.stringify(schemaStructure));
  return hash.digest("hex");
}
