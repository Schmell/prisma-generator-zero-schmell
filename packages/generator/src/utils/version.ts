import { readFile } from "fs/promises";
import { join } from "path";
import { SchemaVersion } from "../types";

export async function getCurrentVersion(
  outputDir: string,
  filename: string
): Promise<SchemaVersion> {
  try {
    const content = await readFile(join(outputDir, filename), "utf-8");
    // Try to match both formats:
    // 1. Old format: version: 1
    // 2. New format: createSchema(1, {
    const versionMatch =
      content.match(/version:\s*(\d+)/) || content.match(/createSchema\(\s*(\d+)/);
    const hashMatch = content.match(/Schema hash: ([a-f0-9]+)/);

    return {
      version: versionMatch ? parseInt(versionMatch[1], 10) : 0,
      hash: hashMatch ? hashMatch[1] : null,
    };
  } catch {
    return { version: 0, hash: null };
  }
}

export function calculateNextVersion(
  currentVersion: number,
  currentHash: string | null,
  newHash: string,
  configVersion?: number
): number {
  if (configVersion !== undefined) {
    return configVersion;
  }

  return currentHash !== newHash ? currentVersion + 1 : currentVersion === 0 ? 1 : currentVersion;
}
