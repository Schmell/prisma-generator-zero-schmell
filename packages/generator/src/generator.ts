import { generatorHandler, GeneratorOptions } from '@prisma/generator-helper'
import { writeFile, mkdir } from 'fs/promises'
import { join } from 'path'
import { transformSchema } from './mappers/schemaMapper'
import { generateCode } from './generators/codeGenerator'
import { getCurrentVersion, calculateNextVersion } from './utils/version'

export async function onGenerate(options: GeneratorOptions) {
  const { generator, dmmf } = options
  const outputFile = 'schema.ts'
  const outputDir = generator.output?.value

  if (!outputDir) {
    throw new Error('Output directory is required')
  }

  const config = {
    name: generator.name,
    prettier: generator.config.prettier === 'true', // Default false,
    resolvePrettierConfig: generator.config.resolvePrettierConfig !== 'false', // Default true
    schemaVersion: generator.config.schemaVersion
      ? Number(generator.config.schemaVersion)
      : undefined,
  }

  // Get current version and hash
  const { version: currentVersion, hash: currentHash } =
    await getCurrentVersion(outputDir, outputFile)

  // Transform the schema
  const transformedSchema = transformSchema(dmmf, currentVersion)

  // Calculate next version
  const nextVersion = calculateNextVersion(
    currentVersion,
    currentHash,
    transformedSchema.hash,
    config.schemaVersion,
  )

  // Update version in transformed schema
  transformedSchema.version = nextVersion

  // Generate code
  let output = generateCode(transformedSchema)

  // Apply prettier if configured
  if (config.prettier) {
    let prettier: typeof import('prettier')
    try {
      prettier = await import('prettier')
    } catch {
      throw new Error('Unable import Prettier. Is it installed?')
    }

    const prettierOptions = config.resolvePrettierConfig
      ? await prettier.resolveConfig(outputFile)
      : null

    output = await prettier.format(output, {
      ...prettierOptions,
      parser: 'typescript',
    })
  }

  await mkdir(outputDir, { recursive: true })
  await writeFile(join(outputDir, outputFile), output)
}

// Use the exported function in the generator handler
generatorHandler({
  onManifest() {
    return {
      version: '1',
      defaultOutput: 'generated/zero',
      prettyName: 'Zero Schema',
    }
  },
  onGenerate,
})
