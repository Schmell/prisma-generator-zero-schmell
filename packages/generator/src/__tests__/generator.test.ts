import { describe, it, expect, beforeEach, vi } from "vitest"
import { DMMF, GeneratorOptions } from "@prisma/generator-helper"
import * as fs from "fs/promises"
import { onGenerate } from "../generator"
import { createField, createModel, createMockDMMF } from "./utils"

// Mock fs/promises
vi.mock("fs/promises", () => ({
  writeFile: vi.fn(),
  mkdir: vi.fn(),
  readFile: vi.fn(),
}))

function createTestOptions(dmmf: DMMF.Document): GeneratorOptions {
  return {
    generator: {
      output: { value: "generated", fromEnvVar: null },
      name: "test-generator",
      config: {},
      provider: { value: "test-provider", fromEnvVar: null },
      binaryTargets: [],
      previewFeatures: [],
      sourceFilePath: "",
    },
    dmmf,
    schemaPath: "",
    datasources: [],
    otherGenerators: [],
    version: "0.0.0",
    datamodel: "",
  }
}

describe("Generator", () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  describe("Schema Generation", () => {
    it("should generate correct schema for basic model", async () => {
      const mockModel: DMMF.Model = {
        name: "User",
        dbName: null,
        fields: [
          createField("id", "String", { isId: true }),
          createField("name", "String"),
          createField("email", "String"),
          createField("age", "Int", { isRequired: false }),
        ],
        uniqueFields: [],
        uniqueIndexes: [],
        primaryKey: null,
      }

      await onGenerate(createTestOptions(createMockDMMF([mockModel])))

      // Verify mkdir was called
      expect(fs.mkdir).toHaveBeenCalledWith("generated", { recursive: true })

      // Verify writeFile was called with correct schema
      const writeFileCalls = vi.mocked(fs.writeFile).mock.calls
      expect(writeFileCalls.length).toBe(1)

      const [, contentBuffer] = writeFileCalls[0]
      const content = contentBuffer.toString()

      expect(content).toMatchSnapshot()
    })

    it("should handle enums correctly", async () => {
      const mockEnum: DMMF.DatamodelEnum = {
        name: "Role",
        values: [
          { name: "USER", dbName: null },
          { name: "ADMIN", dbName: null },
        ],
        dbName: null,
      }

      const mockModel: DMMF.Model = {
        name: "User",
        dbName: null,
        fields: [
          createField("id", "String", { isId: true }),
          createField("role", "Role", { kind: "enum" }),
        ],
        uniqueFields: [],
        uniqueIndexes: [],
        primaryKey: null,
      }

      await onGenerate(
        createTestOptions(createMockDMMF([mockModel], [mockEnum])),
      )

      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[0]
      const content = contentBuffer.toString()

      expect(content).toMatchSnapshot()
    })

    it("should handle relationships correctly", async () => {
      // Create User model with a one-to-many relationship to Post
      const userModel = createModel("User", [
        createField("id", "String", { isId: true }),
        createField("name", "String"),
        createField("posts", "Post", {
          kind: "object",
          isList: true,
          relationName: "UserPosts",
          relationToFields: ["id"],
          relationFromFields: ["userId"],
        }),
      ])

      // Create Post model with both the foreign key and the relation field
      const postModel = createModel("Post", [
        createField("id", "String", { isId: true }),
        createField("title", "String"),
        createField("userId", "String"),
        createField("user", "User", {
          kind: "object",
          relationName: "UserPosts",
          relationFromFields: ["userId"],
          relationToFields: ["id"],
        }),
      ])

      await onGenerate(
        createTestOptions(createMockDMMF([userModel, postModel])),
      )

      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[0]
      const content = contentBuffer.toString()

      // Verify the generated code contains the relationship definitions
      expect(content).toMatchSnapshot()
    })
  })

  describe("Schema Version Management", () => {
    it("should increment version when schema hash changes", async () => {
      // First generation with simpler model creation
      const initialModel = createModel("User", [
        createField("id", "String", { isId: true }),
        createField("name", "String"),
      ])

      // Mock initial file read to simulate no existing file
      vi.mocked(fs.readFile).mockRejectedValueOnce(new Error("File not found"))

      await onGenerate(createTestOptions(createMockDMMF([initialModel])))

      // Get the first generated schema
      const [, firstContentBuffer] = vi.mocked(fs.writeFile).mock.calls[0]
      const firstContent = firstContentBuffer.toString()

      // Second generation with modified model
      const modifiedModel = createModel("User", [
        createField("id", "String", { isId: true }),
        createField("name", "String"),
        createField("email", "String"), // Added field
      ])

      // Mock reading the generated file to be version 1
      vi.mocked(fs.readFile).mockResolvedValueOnce(firstContent)

      await onGenerate(createTestOptions(createMockDMMF([modifiedModel])))

      // Get the content of the last write
      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[1]
      const content = contentBuffer.toString()

      // Version should be incremented
      expect(content).toContain("2,")
    })

    it("should not increment version when schema hash remains the same", async () => {
      const model = createModel("User", [
        createField("id", "String", { isId: true }),
        createField("name", "String"),
      ])

      const dmmf = createMockDMMF([model])

      // Mock reading existing schema with version 1
      vi.mocked(fs.readFile).mockResolvedValueOnce(
        Buffer.from("version: 1\n// Schema hash: abc123"),
      )

      await onGenerate(createTestOptions(dmmf))

      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[
        vi.mocked(fs.writeFile).mock.calls.length - 1
      ]
      const content = contentBuffer.toString()

      // Version should remain the same
      expect(content).toContain("1,")
    })

    it("should use provided schemaVersion from generator config", async () => {
      const model = createModel("User", [
        createField("id", "String", { isId: true }),
        createField("name", "String"),
      ])

      const options = createTestOptions(createMockDMMF([model]))
      // Set the schema version in generator config
      options.generator.config.schemaVersion = "42"

      await onGenerate(options)

      const [, contentBuffer] = vi.mocked(fs.writeFile).mock.calls[
        vi.mocked(fs.writeFile).mock.calls.length - 1
      ]
      const content = contentBuffer.toString()

      // Version should match the provided config value
      expect(content).toContain("42,")
    })
  })
})
