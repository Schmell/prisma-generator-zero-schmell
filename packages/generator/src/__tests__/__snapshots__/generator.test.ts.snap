// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generator > Schema Generation > should generate correct schema for basic model 1`] = `
"// Generated by Zero Schema Generator

import {
  table,
  string,
  boolean,
  number,
  json,
  enumeration,
  relationships,
  createSchema,
  Row,
} from "@rocicorp/zero";

// Define tables

export const userTable = table("User")
  .columns({
    id: string(),
    name: string(),
    email: string(),
    age: number().optional(),
  })
  .primaryKey("id");

// Define schema

export const schema = createSchema(
  1,
  {
    tables: [
      userTable,
    ],
  }
);

// Define types
export type Schema = typeof schema;
export type User = Row<typeof schema.tables.User>;

// DO NOT TOUCH THIS. The schema hash is used to determine if the schema has changed and correctly update the version.
// Schema hash: 0010b112e23e9471e565616609729ed91148ba7141e1232388976164dd1dec26
"
`;

exports[`Generator > Schema Generation > should handle enums correctly 1`] = `
"// Generated by Zero Schema Generator

import {
  table,
  string,
  boolean,
  number,
  json,
  enumeration,
  relationships,
  createSchema,
  Row,
} from "@rocicorp/zero";

// Define enums

export enum Role {
  USER = "USER",
  ADMIN = "ADMIN",
}

// Define tables

export const userTable = table("User")
  .columns({
    id: string(),
    role: enumeration<Role>(),
  })
  .primaryKey("id");

// Define schema

export const schema = createSchema(
  1,
  {
    tables: [
      userTable,
    ],
  }
);

// Define types
export type Schema = typeof schema;
export type User = Row<typeof schema.tables.User>;

// DO NOT TOUCH THIS. The schema hash is used to determine if the schema has changed and correctly update the version.
// Schema hash: b9a9a68f5a33016c062f5b4984d17cb4147f177ccf6a9be1c7dcf49bff9f69b6
"
`;

exports[`Generator > Schema Generation > should handle relationships correctly 1`] = `
"// Generated by Zero Schema Generator

import {
  table,
  string,
  boolean,
  number,
  json,
  enumeration,
  relationships,
  createSchema,
  Row,
} from "@rocicorp/zero";

// Define tables

export const userTable = table("User")
  .columns({
    id: string(),
    name: string(),
  })
  .primaryKey("id");

export const postTable = table("Post")
  .columns({
    id: string(),
    title: string(),
    userId: string(),
  })
  .primaryKey("id");


// Define relationships

export const userTableRelationships = relationships(userTable, ({ many }) => ({
  posts: many({
    sourceField: ["id"],
    destField: ["userId"],
    destSchema: postTable,
  })
}));


export const postTableRelationships = relationships(postTable, ({ one }) => ({
  user: one({
    sourceField: ["userId"],
    destField: ["id"],
    destSchema: userTable,
  })
}));

// Define schema

export const schema = createSchema(
  1,
  {
    tables: [
      userTable,
      postTable,
    ],
    relationships: [
      userTableRelationships,
      postTableRelationships,
    ],
  }
);

// Define types
export type Schema = typeof schema;
export type User = Row<typeof schema.tables.User>;
export type Post = Row<typeof schema.tables.Post>;

// DO NOT TOUCH THIS. The schema hash is used to determine if the schema has changed and correctly update the version.
// Schema hash: e8145ed27becf8118560ddf0a7d341f3d2cf3bf785bbea98a670ba3da4035b4b
"
`;
